hedera-node/docs/design/services/smart-contract-service/system-contract-abi-guidelines.md

```diff
@@ -0,0 +1,166 @@
+  a response code return no output (0 bytes returned) on failure.
+  - The view methods that don’t return a response code should be only those for which invalid
+    input arguments or invalid processing are easy - trivial! - for the developer to figure out
+    *without* a response code.  Any failure that could be attributed to input arguments being
+    Hedera entities (accounts, tokens, contracts, etc.) but having incorrect *properties* for the
+    method to succeed should return a response code so that the particular semantic problem can be
+    highlighted.
+    - Avoid inventing new response codes if there's an existing one that will serve. New response
+      codes must be mentioned in the related HIP.
+    - An example where invalid arguments are trivial to figure out and the method could _not_ return
+      a response code would be `isValidAlias(address)` - it either is a valid alias or it isn't.
+      There's nothing more to be said about it (and in fact there's nothing more that _could_ be
+      said about it).
+- System contract methods should not revert - instead signal failure via the response code (as above)
+  - Reverting messes up calling contracts.
+- Protobuf encoding of arguments must only be used when the only use case is for the input argument
+  to be provided from an external source (e.g., a DApp) and it is to be passed through the called
+  contract into the system contract method without needing interpretation or alteration (same for
+  returns).  Our APIs must not use a protobuf-encoded argument if there's a reasonable use case that
+  calls for a (user's) contract to serialize into a protobuf and then call a system contract method.
+  - There is no *native* support of protobufs in Solidity.  There is limited support for decoding
+    or encoding protobufs via open-source libraries:
+    - Limited: There’s a reasonable open source protobuf codec ([`protobuf3-solidity-lib`](https://github.com/celestiaorg/protobuf3-solidity-lib))
+      and `protoc` plugin ([`protobuf3-solidity`](https://github.com/celestiaorg/protobuf3-solidity)),
+      both from [Celestia](https://celestia.org); and then Solidity itself limits how much of the
+      protobuf spec that codec/plugin supports.
+      - If you *intend* to use this codec/plugin you must design your protobufs accordingly.
+        (As an example, our existing HAPI protobufs already use features that are unavailable
+        for this codec.)
+        - Significant documented limitations (at this time) are: no repeated `string` or `bytes`,
+          no `oneof`, no `map`s, and no nested `enum` or `message`.
+  - Example of it used this way: HIP-632 (Hedera Account Service) `isAuthorized` method where the
+    smart contract caller packages a set of public key/signature pairs into a protobuf and passes
+    that into its smart contract which just passes it, uninspected and untouched, into the system
+    via `isAuthorized`.
```

Comment:
question: does this mean that services is unmarshalling a user encoded protobuf? could this lead to problems if the protobuf is malicious?

Reply:
It does mean that and one of the reasons we use PBJ (i.e., our own protobuf compiler) is to protect against that: It's a key requirement of PBJ. (And after all, _all_ HAPI calls are protobufs that could be malicious.)